text/000-rustdoc-cfgs-handling.md

```diff
@@ -0,0 +1,283 @@
+
+  * `#[doc(auto_cfg)]`/`#[doc(no_auto_cfg)]`
+
+    When this is turned on, `#[cfg]` attributes are shown in documentation just like `#[doc(cfg)]` attributes are.
+
+  * `#[doc(cfg_hide(...))]` / `#[doc(cfg_show(...))]`
+
+    These attributes suppress or un-suppress the `auto_cfg` behavior for a particular configuration predicate.
+
+    For example, `#[doc(cfg_hide(windows))]` shall be used in newer versions of the [`windows` crate] to prevent the "this is supported on **windows** only" tag from being shown on every single item.
+
+[cfg attribute]: https://doc.rust-lang.org/reference/conditional-compilation.html
+[`windows` crate]: https://docs.rs/windows/latest/windows/
+
+All of these attributes can be added to a module or to the crate root, and they will be inherited by the child items unless another attribute overrides it (except that `doc(cfg)` cannot be added to the crate root). This is why "opposite" attributes like `cfg_hide` and `cfg_show` are provided: they allow a child item to override its parent.
+
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## The attributes
+
+### `#[doc(cfg(...))]`
+
+This attribute provides a standardized format to document conditionally available items. Example:
+
+```rust
+// the "real" cfg condition
+#[cfg(feature = "futures-io")]
+// the `doc(cfg())` so it's displayed to the readers
+#[doc(cfg(feature = "futures-io"))]
+pub mod futures {}
+```
+
+It will display in the documentation for this module:
+
+![This is supported on feature="futures-io" only.](https://user-images.githubusercontent.com/81079/89731116-d7b7ce00-da44-11ea-87c6-022d192d6eca.png)
+
+This attribute has the same syntax as conditional compilation, but it only causes documentation to be added. This means `#[doc(cfg(false))` will not cause your docs to be hidden, even though `#[cfg(false)]` does do that.
+
+This attribute works on modules and on items but cannot be used at the crate root level.
+
+### `#[doc(auto_cfg)]`/`#[doc(no_auto_cfg)]`
+
+By default, `#[doc(auto_cfg)]` is enabled at the crate-level. When it's enabled, Rustdoc will automatically display `cfg(...)` compatibility information as-if the same `#[doc(cfg(...))]` had been specified.
+
+So if we take back the previous example:
+
+```rust
+#[cfg(feature = "futures-io")]
+pub mod futures {}
+```
+
+There's no need to "duplicate" the `cfg` into a `doc(cfg())` to make Rustdoc display it.
+
+In some situations, the detailed conditional compilation rules used to implement the feature might not serve as good documentation (for example, the list of supported platforms might be very long, and it might be better to document them in one place). To turn it off, add the `#[doc(no_auto_cfg)]` attribute.
+
+Both `#[doc(auto_cfg)]` and `#[doc(no_auto_cfg)]` attributes impact all there descendants. You can then enable/disable them by using the opposite attribute on a given item. They can be used as follows:
```

Comment:
The current implementation only allows these attributes at the crate-level, has there been a push to support this arbitrary nesting from somewhere?

Reply:
(Well, actually the attributes don't exist, `feature(doc_auto_cfg)` _implies_ default-active `doc(auto_cfg)`, but IIRC previous discussions about how to expose a stable toggle had only considered adding in the attribute toggling it at the crate-level).

Reply:
> The current implementation only allows these attributes at the crate-level, has there been a push to support this arbitrary nesting from somewhere?

No, it's a change brought by this RFC: users will be able to use them on items directly if they want to have better control where they want it to be enabled or not.

Just like lints, I think it's better to give better control over this. For example if someone wants to completely override doc cfg attributes on a given item, they will be able to disable it for this item completely and then add what they want using `doc(cfg())`. For this I'm thinking mostly about if you have "parent" features with a lot of children and for some reasons you have a complex `cfg` to handle it and want to simplify its display when rendered in documentation.

Reply:
`#[doc(cfg(...))]` already overrides `#[cfg(...)]` on a single item, you don't need to disable the auto-cfg for that.

The use case where you would want to disable `doc(auto_cfg)` on a sub-tree seems to be if you have a lot of `cfg` on sub-items of that tree that you don't want rendered at all (but these `cfg` _aren't_ ones you want to globally `#[cfg_hide(...)]`), and so by disabling it you get to avoid putting `#[doc(cfg())]` on each sub-item individually. But that seems like an unlikely situation to me.

Similarly wanting to `#[cfg_hide(...)]` some `cfg` within just a sub-tree seems very unlikely,

Making these attributes work within the module tree brings in extra complexity by requiring the inverses to exist. If they're restricted to applying only at the crate root we only need two new attributes: `#[doc(no_auto_cfg)]` (assuming the default is active) and `#[doc(cfg_hide(...))]`.

Reply:
Fair point. I'll limit `doc(auto_cfg)` as a crate-level attribute.

For `cfg_hide` I'm less convinced as you might want to hide some cfgs only for some modules/items (this is very hypothetical though). As for the complexity, it seems pretty low in a visitor: we add new entries before iterating to the children and remove them once we leave the current item (but that's just an implementation detail, not sure it's something that we should take into account here).

Anyway, if no one seems to be interested into having `cfg_show`/`cfg_hide` to be used on a specific module/item, then I will turn them into crate-level attribute (well, `cfg_show` would be useless in this case so it'll be simply removed).

Reply:
I just thought about a use case where having `doc(auto_cfg)` not only as a crate-level attribute would be a nice improvement! In sysinfo, I disabled the `doc_auto_cfg` because every item is available, whatever the OS (even unsupported ones). Except that I recently splitted the components to enable them through features. So I can either do:

```rust
#![doc(auto_cfg = true)]
#![doc(cfg_hide(/* list of all supported OSes */)]
```

Or instead disable `auto_cfg` at the crate level and enable it on the components module directly without needing to worry about hiding `cfg`s since the OS information handling is one level upper.

So with this in mind, I think we should lift the crate-level only restriction on `auto_cfg`. What do you think?